internal/provider/resource_gitlab_tag.go

```diff
@@ -0,0 +1,171 @@
+	return &schema.Resource{
+		Description: "This resource allows you to create and manage GitLab tags.",
+
+		CreateContext: resourceGitlabTagCreate,
+		ReadContext:   resourceGitlabTagRead,
+		DeleteContext: resourceGitlabTagDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "The name of a tag.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+			},
+			"project": {
+				Description: "The ID or URL-encoded path of the project owned by the authenticated user.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+			},
+			"ref": {
+				Description: "Create tag using commit SHA, another tag name, or branch name.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+			},
+			"message": {
+				Description: "Creates annotated tag.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+			},
+			"protected": {
+				Description: "Bool, true if tag has tag protection.",
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+			"target": {
+				Description: "The unique id assigned to the commit by Gitlab.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"release": {
+				Description: "The release associated with the tag.",
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Set:         schema.HashResource(releaseNoteSchema),
+				Elem:        releaseNoteSchema,
+			},
+			"commit": {
+				Description: "The commit associated with the tag.",
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Set:         schema.HashResource(commitSchema),
+				Elem:        commitSchema,
+			},
+		},
+	}
+})
+
+var releaseNoteSchema = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"tag_name": {
+			Description: "The name of the tag.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"description": {
+			Description: "The description of release.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+	},
+}
+
+func resourceGitlabTagCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	name := d.Get("name").(string)
+	project := d.Get("project").(string)
+	ref := d.Get("ref").(string)
+	message := d.Get("message").(string)
+	tagOptions := &gitlab.CreateTagOptions{
+		TagName: &name, Ref: &ref, Message: &message,
+	}
+
+	log.Printf("[DEBUG] create gitlab tag %s/%s with ref %s", project, name, ref)
+	_, resp, err := client.Tags.CreateTag(project, tagOptions, gitlab.WithContext(ctx))
+	if err != nil {
+		log.Printf("[DEBUG] failed to create gitlab tag %s/%s response %v", project, name, resp)
+		return diag.FromErr(err)
+	}
+	d.SetId(buildTwoPartID(&project, &name))
+	d.Set("ref", ref)
+	d.Set("message", message)
```

Comment:
The message will always be set in the read function, right? 

```suggestion
```

Reply:
Pardon me. I had skipped this moment)